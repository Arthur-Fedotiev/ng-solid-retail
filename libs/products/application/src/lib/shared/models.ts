import { Categories, Colors, Retailers, Tiers } from './constants';

export type CategoriesMap = typeof Categories;
export type Category = typeof Categories[keyof typeof Categories];

export type RetailersMap = typeof Retailers;
export type TiersMap = typeof Tiers;
export type ColorsMap = typeof Colors;
