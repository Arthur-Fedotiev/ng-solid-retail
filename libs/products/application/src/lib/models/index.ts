export * from './category.view-model';
export * from './create-product-from.interface';
export * from './product.view-model';
export * from './products-state.model';
export * from './retailer.view-model';
export * from './product-by-price.type';
export * from './price.view-model';
