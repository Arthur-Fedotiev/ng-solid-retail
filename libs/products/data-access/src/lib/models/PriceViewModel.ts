import { Price } from '@omnia/products/domain';
import { RetailerViewModel } from './RetailerViewModel';

export class PriceViewModel {
  constructor(
    dto: Price,
    public readonly id: string = dto.id,
    public readonly productId: string = dto.productId,
    public readonly retailer: RetailerViewModel = new RetailerViewModel(
      dto.Retailer.id,
      dto.Retailer.Name
    ),
    public readonly tier: number = dto.Tier,
    public readonly updateTime: string = dto.UpdateTime,
    public readonly price: number = Number(dto.Price)
  ) {}
}
