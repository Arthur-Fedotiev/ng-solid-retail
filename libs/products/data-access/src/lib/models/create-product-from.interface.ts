export interface CreateProductForm {
  name: string;
  sku: string;
  description: string;

  categories: ReadonlyArray<{ id: string; name: string }>;
  prices: ReadonlyArray<{
    price: number;
    tier: number;
    retailer: { id: string; name: string };
  }>;
}
