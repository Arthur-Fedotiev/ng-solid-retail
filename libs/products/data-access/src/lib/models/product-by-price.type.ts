import { ProductViewModel } from '../..';

export type ProductByPrice = Pick<ProductViewModel, 'id' | 'name' | 'sku'> & {
  price: { price: number; retailer: { name: string } };
};
