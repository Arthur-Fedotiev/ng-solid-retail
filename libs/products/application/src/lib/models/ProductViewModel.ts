import { Product } from '@sr/products/entities';
import { CategoryEnum } from '../constants/category.enum';
import { PriceViewModel } from './PriceViewModel';
import { CategoryViewModel } from './CategoryViewModel';

export class ProductViewModel {
  constructor(
    private readonly dto: Product,
    public readonly id: string = dto.id,
    public readonly name: string = dto.Name,
    public readonly description: string = dto.Description,
    public readonly sku: string = dto.SKU,
    public readonly url: string = dto.Url,
    public readonly categories: ReadonlyArray<CategoryViewModel> = dto.Categories.map(
      ({ id, Name }) =>
        new CategoryViewModel(id as string, Name as CategoryEnum)
    ),
    public readonly prices: ReadonlyArray<PriceViewModel> = dto.Prices.map(
      (price) => new PriceViewModel(price)
    )
  ) {}

  public clone(dto: Partial<ProductViewModel> = {}): ProductViewModel {
    return Object.assign(new ProductViewModel(this.dto), dto);
  }
}
