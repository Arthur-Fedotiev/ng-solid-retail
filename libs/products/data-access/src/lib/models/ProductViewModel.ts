import { Product } from '@omnia/products/domain';
import { CategoryEnum } from '../constants/category.enum';
import { PriceViewModel } from './PriceViewModel';
import { CategoryViewModel } from './CategoryViewModel';

export class ProductViewModel {
  constructor(
    dto: Product,
    public readonly id: string = dto.id,
    public readonly name: string = dto.Name,
    public readonly description: string = dto.Description,
    public readonly sku: string = dto.SKU,
    public readonly categories: ReadonlyArray<CategoryViewModel> = dto.Categories.map(
      ({ id, Name }) =>
        new CategoryViewModel(id as string, Name as CategoryEnum)
    ),
    public readonly prices: ReadonlyArray<PriceViewModel> = dto.Prices.map(
      (price) => new PriceViewModel(price)
    )
  ) {}
}
