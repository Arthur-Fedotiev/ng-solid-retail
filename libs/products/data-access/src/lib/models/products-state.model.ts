import { CategoryViewModel } from './CategoryViewModel';
import { ProductViewModel } from './ProductViewModel';
import { RetailerViewModel } from './RetailerViewModel';

export class ProductsStateModel {
  public products: ReadonlyArray<ProductViewModel> | null;
  public readonly categories: ReadonlyArray<CategoryViewModel> | null;
  public readonly retailers: ReadonlyArray<RetailerViewModel> | null;

  public readonly selectedProduct: ProductViewModel | null = null;
  public readonly loading = false;
  public readonly error: string | null = null;

  constructor(
    products?: ReadonlyArray<ProductViewModel>,
    categories?: ReadonlyArray<CategoryViewModel>,
    retailers?: ReadonlyArray<RetailerViewModel>
  ) {
    this.products = products ? products : null;
    this.categories = categories ? categories : null;
    this.retailers = retailers ? retailers : null;
  }
}
