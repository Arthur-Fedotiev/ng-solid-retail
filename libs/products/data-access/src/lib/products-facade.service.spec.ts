import { fakeAsync, TestBed, tick } from '@angular/core/testing';
import { Product, ProductsApi } from '@omnia/products/domain';
import { PRODUCTS_API, makeProductsStub } from '@omnia/products/infrastructure';
import { ID_GENERATOR } from '@omnia/shared/util';
import { asyncScheduler, of, scheduled } from 'rxjs';
import { ProductsStateModel } from './models/products-state.model';
import { ProductViewModel } from './models/ProductViewModel';

import { ProductsFacadeService } from './products-facade.service';
import { TO_PRODUCT_POST_DTO } from './providers/to-product-post-dto.token';
import { makeProductViewModelsStub } from './testing/make-product-view-models-stub';

describe('ProductsFacadeService', () => {
  let service: ProductsFacadeService;
  let productsApiProviderMock: jest.Mocked<ProductsApi>;
  let toProductPostDtoProviderMock: jest.Mock;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [
        {
          provide: PRODUCTS_API,
          useValue: {
            getProducts: jest.fn(),
            getCategories: jest.fn(),
            createProduct: jest.fn(),
            getRetailers: jest.fn(),
            deleteProduct: jest.fn().mockReturnValue(of()),
          },
        },
        {
          provide: TO_PRODUCT_POST_DTO,
          useValue: jest.fn(),
        },
        {
          provide: ID_GENERATOR,
          useValue: () => 'id',
        },
      ],
    });
    service = TestBed.inject(ProductsFacadeService);
    productsApiProviderMock = TestBed.inject(
      PRODUCTS_API
    ) as jest.Mocked<ProductsApi>;
    toProductPostDtoProviderMock = TestBed.inject(
      TO_PRODUCT_POST_DTO
    ) as jest.Mock;
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  afterEach(() => jest.clearAllMocks());

  describe('loadProducts', () => {
    let productsStub: ReadonlyArray<Product>;

    beforeEach(() => (productsStub = makeProductsStub(5)));

    it('should delegate to GetProducts', fakeAsync(() => {
      productsApiProviderMock.getProducts.mockReturnValue(
        scheduled([productsStub], asyncScheduler)
      );

      service.loadProducts();
      tick();

      expect(productsApiProviderMock.getProducts).toHaveBeenCalledTimes(1);
    }));

    it('should set product state', fakeAsync(() => {
      const expected = productsStub.map((p) => new ProductViewModel(p));

      productsApiProviderMock.getProducts.mockReturnValue(
        scheduled([productsStub], asyncScheduler)
      );

      service.loadProducts();

      service.products$.subscribe((products) => {
        expect(products).toEqual(expected);
      });

      tick();
    }));
  });

  describe('products state derivations', () => {
    it('should derive productsList for each product price', fakeAsync(() => {
      const productsViewModelsStub = makeProductViewModelsStub(1);
      const expected = productsViewModelsStub.reduce(
        (count, product) => count + product.prices.length,
        0
      );

      service.state$.next(new ProductsStateModel(productsViewModelsStub));

      tick();

      service.productsListForEachPrice$.subscribe((products) => {
        expect(products.length).toEqual(expected);
      });

      tick();
    }));

    it('should derive products short info', fakeAsync(() => {
      const productsViewModelsStub = makeProductViewModelsStub(1);
      const expectedShortInfo = {
        id: productsViewModelsStub[0].id,
        price: productsViewModelsStub[0].prices[0].price,
        name: productsViewModelsStub[0].name,
        sku: productsViewModelsStub[0].sku,
        retailer: productsViewModelsStub[0].prices[0].retailer.name,
      };

      service.state$.next(new ProductsStateModel(productsViewModelsStub));

      tick();

      service.productsShortInfo$.subscribe((products) => {
        expect(products[0]).toEqual(expectedShortInfo);
      });

      tick();
    }));
  });

  describe('loadCategories', () => {
    it('should delegate to GetCategories', fakeAsync(() => {
      productsApiProviderMock.getCategories.mockReturnValue(
        scheduled([[]], asyncScheduler)
      );

      service.loadCategories();
      tick();

      expect(productsApiProviderMock.getCategories).toHaveBeenCalledTimes(1);
    }));

    it('should set categories state', fakeAsync(() => {
      const categoriesStub = [
        { id: '1', Name: 'Category 1' },
        { id: '2', Name: 'Category 2' },
      ];

      const expected = categoriesStub.map((c) => ({ id: c.id, name: c.Name }));

      productsApiProviderMock.getCategories.mockReturnValue(
        scheduled([categoriesStub], asyncScheduler)
      );

      service.loadCategories();

      service.categories$.subscribe((categories) => {
        expect(categories).toEqual(expected);
      });

      tick();
    }));
  });

  describe('#createProduct', () => {
    xit('should delegate to CreateProduct passing  POST dto', fakeAsync(() => {
      const product = makeProductViewModelsStub(1)[0];

      toProductPostDtoProviderMock.mockReturnValue(product);

      service.createProduct(product);

      expect(productsApiProviderMock.createProduct).toHaveBeenCalledWith(
        product
      );
    }));
  });

  describe('#loadRetailers', () => {
    it('should delegate to GetRetailers', fakeAsync(() => {
      productsApiProviderMock.getRetailers.mockReturnValue(
        scheduled([[]], asyncScheduler)
      );

      service.loadRetailers();
      tick();

      expect(productsApiProviderMock.getRetailers).toHaveBeenCalledTimes(1);
    }));

    it('should set retailers state', fakeAsync(() => {
      const retailersStub = [
        { id: '1', Name: 'Retailer 1' },
        { id: '2', Name: 'Retailer 2' },
      ];

      const expected = retailersStub.map((r) => ({ id: r.id, name: r.Name }));

      productsApiProviderMock.getRetailers.mockReturnValue(
        scheduled([retailersStub], asyncScheduler)
      );

      service.loadRetailers();

      tick();

      service.retailers$.subscribe((retailers) => {
        expect(retailers?.length).toEqual(expected.length);
      });

      tick();
    }));
  });

  describe('#deleteProduct', () => {
    it('should delegate to DeleteProduct passing id', fakeAsync(() => {
      const id = 'id';

      productsApiProviderMock.deleteProduct.mockReturnValue(of());

      service.deleteProduct(id);

      expect(productsApiProviderMock.deleteProduct).toHaveBeenCalledWith(id);
    }));

    it('should remove product from state', fakeAsync(() => {
      const productsStub = makeProductViewModelsStub(1);
      const id = productsStub[0].id;

      Object.assign(service, { state: new ProductsStateModel(productsStub) });
      productsApiProviderMock.deleteProduct.mockReturnValue(of());

      service.deleteProduct(id);

      service.products$.subscribe((products) => {
        expect(products.length).toEqual(0);
      });

      tick();
    }));
  });
});
