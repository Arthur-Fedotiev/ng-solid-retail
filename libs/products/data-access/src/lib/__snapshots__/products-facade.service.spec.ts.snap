// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsFacadeService #updateProductPrice should delegate to UpdateProductPrice passing updated product an updated price 1`] = `
Array [
  Object {
    "Prices": Array [
      Object {
        "id": "1",
      },
    ],
  },
  Object {
    "id": "1",
  },
]
`;
